testing/spectypes/premine-state.go

```diff
@@ -0,0 +1,426 @@
+	if err := g.SetRandaoMixes(rao); err != nil {
+		return err
+	}
+	if err := g.SetBlockRoots(nZeroRoots(uint64(params.BeaconConfig().SlotsPerHistoricalRoot))); err != nil {
+		return err
+	}
+	if err := g.SetStateRoots(nZeroRoots(uint64(params.BeaconConfig().SlotsPerHistoricalRoot))); err != nil {
+		return err
+	}
+	if err := g.SetSlashings(make([]uint64, params.BeaconConfig().EpochsPerSlashingsVector)); err != nil {
+		return err
+	}
+	if err := s.setLatestBlockHeader(g); err != nil {
+		return err
+	}
+	if err := s.setInactivityScores(g); err != nil {
+		return err
+	}
+	if err := s.setSyncCommittees(g); err != nil {
+		return err
+	}
+	if err := s.setExecutionPayload(g); err != nil {
+		return err
+	}
+
+	// For pre-mined genesis, we want to keep the deposit root set to the root of an empty trie.
+	// This needs to be set again because the methods used by processDeposits mutate the state's eth1data.
+	if err := s.setEth1Data(g); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *premineGenesisConfig) setGenesisValidatorsRoot(g state.BeaconState) error {
+	vroot, err := stateutil.ValidatorRegistryRoot(g.Validators())
+	if err != nil {
+		return err
+	}
+	return g.SetGenesisValidatorsRoot(vroot[:])
+}
+
+func (s *premineGenesisConfig) setFork(g state.BeaconState) error {
+	var pv, cv []byte
+	switch s.Version {
+	case version.Phase0:
+		pv, cv = params.BeaconConfig().GenesisForkVersion, params.BeaconConfig().GenesisForkVersion
+	case version.Altair:
+		pv, cv = params.BeaconConfig().GenesisForkVersion, params.BeaconConfig().AltairForkVersion
+	case version.Bellatrix:
+		pv, cv = params.BeaconConfig().AltairForkVersion, params.BeaconConfig().BellatrixForkVersion
+	default:
+		return errUnsupportedVersion
+	}
+	fork := &ethpb.Fork{
+		PreviousVersion: pv,
+		CurrentVersion:  cv,
+		Epoch:           0,
+	}
+	return g.SetFork(fork)
+}
+
+func (s *premineGenesisConfig) setInactivityScores(g state.BeaconState) error {
+	if s.Version < version.Altair {
+		return nil
+	}
+
+	scores, err := g.InactivityScores()
+	if err != nil {
+		return err
+	}
+	scoresMissing := len(g.Validators()) - len(scores)
+	if scoresMissing > 0 {
+		for i := 0; i < scoresMissing; i++ {
+			scores = append(scores, 0)
+		}
+	}
+	return g.SetInactivityScores(scores)
+}
+
+func (s *premineGenesisConfig) setSyncCommittees(g state.BeaconState) error {
+	if s.Version < version.Altair {
+		return nil
+	}
+	sc, err := altair.NextSyncCommittee(context.Background(), g)
+	if err != nil {
+		return err
+	}
+	if err = g.SetNextSyncCommittee(sc); err != nil {
+		return err
+	}
+	if err = g.SetCurrentSyncCommittee(sc); err != nil {
+		return err
+	}
+	return nil
+}
+
+type rooter interface {
+	HashTreeRoot() ([32]byte, error)
+}
+
+func (s *premineGenesisConfig) setLatestBlockHeader(g state.BeaconState) error {
+	var body rooter
+	switch s.Version {
+	case version.Phase0:
+		body = &ethpb.BeaconBlockBody{
```

Comment:
maybe it would be nice to factor this into a method that `NewGenesisBlockForState` could share. Note the only difference is that `NewGenesisBlockForState` wants to actually set the `StateRoot` attribute in the body, but it could do that after calling the shared method. I didn't look to see if the beacon block interface has comprehensive setters like BeaconState does. If so I could rewrite this to use the interface setters as well.